### PR TITLE
add support of emmeans to bootstrapped models

### PR DESCRIPTION
**Pull request by mattansb**
This PR adds emmeans-methods for the bootstrap functions (see examples).

Some notes:
1. To do this I have to save the model and the bootstrapped estimates in the output object from both `bootstrap_model()` and `bootstrap_parameters()`. @strengejacke is that okay? 
2. This can probably be extended to the `simulate_model()` and `simulate_parameters()` as well.
3. (Wasn't sure where to address this in the `NEWS` file)

``` r
library(parameters)
library(emmeans)

model <- lm(mpg ~ log(wt) + factor(cyl), data = mtcars)

b <- bootstrap_model(model, iterations = 50)
head(b)
#>   (Intercept)    log(wt) factor(cyl)6 factor(cyl)8
#> 1    35.46647 -10.835075    -3.336407    -5.259839
#> 2    37.01599 -12.766090    -2.532144    -4.536621
#> 3    34.41191  -9.417361    -3.953449    -6.906399
#> 4    36.23839  -9.453581    -5.618538    -8.500304
#> 5    35.95811 -11.459697    -3.400162    -5.137954
#> 6    33.72223 -10.223921    -2.666147    -4.839145

emmeans(b, consec ~ cyl)
#> $emmeans
#>  cyl emmean lower.HPD upper.HPD
#>    4   22.5      19.8      26.5
#>    6   19.3      18.7      20.1
#>    8   17.4      16.2      18.9
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 
#> 
#> $contrasts
#>  contrast estimate lower.HPD upper.HPD
#>  6 - 4       -3.14     -7.00    -0.675
#>  8 - 6       -2.01     -3.17     0.330
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95
```

or

``` r
(b <- bootstrap_parameters(model, iterations = 50))
#> Parameter    | Coefficient |          95% CI |      p
#> -----------------------------------------------------
#> (Intercept)  |       35.98 | [ 33.07, 39.12] | < .001
#> log(wt)      |      -11.39 | [-14.49, -8.43] | < .001
#> factor(cyl)6 |       -3.45 | [ -5.14, -0.75] | < .001
#> factor(cyl)8 |       -5.00 | [ -7.19, -2.30] | < .001

emmeans(b, consec ~ cyl)
#> $emmeans
#>  cyl emmean lower.HPD upper.HPD
#>    4   22.7      19.6      24.4
#>    6   19.3      18.3      20.1
#>    8   17.5      15.8      18.3
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 
#> 
#> $contrasts
#>  contrast estimate lower.HPD upper.HPD
#>  6 - 4       -3.45     -5.22    -0.313
#>  8 - 6       -1.85     -3.23    -0.209
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95
```

<sup>Created on 2021-01-31 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

File: R/bootstrap_model-emmeans.R [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#' @keywords emmeans_methods
+emm_basis.bootstrap_model <- function(object, trms, xlev, grid, ...) {
+  if (!requireNamespace("emmeans", quietly = TRUE))
+    stop("Thin function requires the 'emmeans' package to work.")
```
Comment (strengejacke) [inline R/bootstrap_model-emmeans.R:4]: typo

File: R/bootstrap_model-emmeans.R [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    stop("Oops! Cannot create the reference grid. Please open an issue at github.com/easystats/parameters/issues.")
+
+  emb$post.beta <- as.matrix(object)
+  emb
+}
+
+#' @keywords emmeans_methods
+recover_data.bootstrap_model <- function(object, ...) {
+  if (!requireNamespace("emmeans", quietly = TRUE))
+    stop("Thin function requires the 'emmeans' package to work.")
```
Comment (strengejacke) [inline R/bootstrap_model-emmeans.R:19]: typo

File: R/bootstrap_model-emmeans.R [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  model <- attr(object, "original_model")
+  emmeans::recover_data(model, ...)
+}
+
+
+#' @keywords emmeans_methods
+emm_basis.bootstrap_parameters <- function(object, trms, xlev, grid, ...) {
+  if (!requireNamespace("emmeans", quietly = TRUE))
+    stop("Thin function requires the 'emmeans' package to work.")
```
Comment (strengejacke) [inline R/bootstrap_model-emmeans.R:29]: typo

File: R/bootstrap_model-emmeans.R [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  model <- attr(object, "boot_samples")
+  emmeans::emm_basis(model, trms, xlev, grid, ...)
+}
+
+#' @keywords emmeans_methods
+recover_data.bootstrap_parameters <- function(object, ...) {
+  if (!requireNamespace("emmeans", quietly = TRUE))
+    stop("Thin function requires the 'emmeans' package to work.")
```
Comment (strengejacke) [inline R/bootstrap_model-emmeans.R:38]: typo

**Review by strengejacke — CHANGES_REQUESTED** [review]

**Comment by strengejacke** [comment]
Looks good to me. Extending to `simulate_parameters()` (or better: `simulate_model()`, `simulate_parameters()` just summarizes `simulate_model()`) would be nice.

**Comment by mattansb** [comment]
Do you have any reservations about having the actual model saved as an attribute? We usually don't do this...

**Comment by strengejacke** [comment]
> Do you have any reservations about having the actual model saved as an attribute? We usually don't do this...

No, I think in this case we can allow this "overload". And I think there are indeed some functions that do this as well, maybe in _performance_? I cannot recall exactly, but I think we had this case already.

**Comment by strengejacke** [comment]
You may look into the checks:
https://github.com/easystats/parameters/pull/411/checks?check_run_id=1800764150

Maybe related to your PR?

File: R/bootstrap_model.R [hunk 5]
```diff
@@ -7,15 +7,25 @@
...
 #' \donttest{
 #' if (require("boot")) {
-#'   model <- lm(mpg ~ wt + cyl, data = mtcars)
-#'   head(bootstrap_model(model))
+#'   model <- lm(mpg ~ wt + factor(cyl), data = mtcars)
+#'   b <- bootstrap_model(model)
+#'   print(head(b))
+#'
+#'   if (require("emmeans")) {
+#'     print(emmeans(b, consec ~ cyl))
```
Comment (strengejacke) [inline R/bootstrap_model.R:27]: So why is it `consec` here?

**Comment by mattansb** [comment]
That code seems to be in `boot`, so unrelated to my changes. Anyway it's wrapped in a `donttest`, so it shouldn't fail on CRAN checks, right?

File: R/bootstrap_model.R (hunk 5 above)
Comment (mattansb) [reply R/bootstrap_model.R:27]: No particular reason - just wanted to show off some other contrasts methods other than `pairwise`...

File: R/bootstrap_model.R [hunk 6]
```diff
@@ -7,15 +7,25 @@
 #' @param ... Arguments passed to or from other methods.
 #' @inheritParams p_value
 #'
-#' @return A data frame.
+#' @return A data frame of bootstrapped estimates.
+#'
+#' @section Using with \code{emmeans}:
+#' The output can be passed directly to the various functions from the
+#' \code{emmeans} package, to obtain bootstrapped estimates, contrasts, siple
```
Comment (strengejacke) [inline R/bootstrap_model.R:14]: typo - "siple"

**Comment by mattansb** [comment]
If all is good, I'll wait to hear from Russel before I merge this.

File: R/bootstrap_model.R (hunk 5 above)
Comment (strengejacke) [reply R/bootstrap_model.R:27]: Ah, ok. I thought that was a variable name...

**Comment by mattansb** [comment]
Ohhh... look at this!

``` r
library(parameters)
library(emmeans)

model <- lm(mpg ~ log(wt) + factor(cyl), data = mtcars)

b <- bootstrap_model(model, iterations = 200)
head(b)
#>   (Intercept)    log(wt) factor(cyl)6 factor(cyl)8
#> 1    33.85733  -9.398716    -4.336239    -5.979486
#> 2    38.39998 -13.585232    -3.739893    -3.898692
#> 3    37.92651 -13.307826    -2.798281    -3.673321
#> 4    33.78963  -8.865101    -3.442550    -6.929030
#> 5    35.83263 -12.261401    -2.458065    -4.127258
#> 6    36.82099 -13.453800    -1.842331    -2.402226

em_ <- emmeans(b, consec ~ cyl)

standard_error(em_)
#>   Parameter        SE Component
#> 1         4 1.2192554   emmeans
#> 2         6 0.4758854   emmeans
#> 3         8 0.6276993   emmeans
#> 4     6 - 4 1.2666211 contrasts
#> 5     8 - 6 0.7851086 contrasts
p_value(em_)
#>   Parameter           p Component
#> 1         4 0.004975124   emmeans
#> 2         6 0.004975124   emmeans
#> 3         8 0.004975124   emmeans
#> 4     6 - 4 0.034825871 contrasts
#> 5     8 - 6 0.024875622 contrasts

model_parameters(em_)
#> # Estimated Marginal Means
#> 
#> Parameter | Coefficient |   SE |         95% CI |     p
#> -------------------------------------------------------
#> 4         |       22.44 | 1.22 | [20.01, 24.71] | 0.005
#> 6         |       19.36 | 0.48 | [18.54, 20.30] | 0.005
#> 8         |       17.30 | 0.63 | [16.32, 18.70] | 0.005
#> 
#> # Contrasts
#> 
#> Parameter | Coefficient |   SE |         95% CI |     p
#> -------------------------------------------------------
#> 6 - 4     |       -3.14 | 1.27 | [-4.98,  0.16] | 0.035
#> 8 - 6     |       -1.94 | 0.79 | [-3.62, -0.64] | 0.025
```

<sup>Created on 2021-02-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


(I also streamlined some of the general emmeans code.)

**Comment by strengejacke** [comment]
See https://github.com/easystats/parameters/pull/411/checks?check_run_id=1813845843

**Comment by mattansb** [comment]
I see only one error, right?

```
Running examples in ‘parameters-Ex.R’ failed
  The error most likely occurred in:
  
  > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
  > ### Name: bootstrap_model
  > ### Title: Model bootstrapping
  > ### Aliases: bootstrap_model
  > 
  > ### ** Examples
  > 
  > ## No test: 
  > if (require("boot")) {
  +   model <- lm(mpg ~ wt + factor(cyl), data = mtcars)
  +   b <- bootstrap_model(model)
  +   print(head(b))
  + 
  +   if (require("emmeans")) {
  +     est <- emmeans(b, consec ~ cyl)
  +     print(model_parameters(est))
  +   }
  + }
  Loading required package: boot
  Error in t.star[r, ] <- res[[r]] : 
    number of items to replace is not a multiple of replacement length
  Calls: bootstrap_model -> bootstrap_model.default -> <Anonymous>
```

This is an error in the `boot::boot` function. It is wrapped in a `\donttest` so this shouldn't be an issue, right?

**Comment by strengejacke** [comment]
haha, the other failing check is from my code:

```r
  ci.varest: no visible binding for global variable ‘model’
  ci.varest : <anonymous>: no visible binding for global variable ‘model’
  Undefined global functions or variables:
```

**Comment by mattansb** [comment]
Is there anything else we need / am I missing?

**Comment by strengejacke** [comment]
I think not.

**Comment by mattansb** [comment]
Then I'll add some tests tomorrow (:

**Pull request merged**